*******************************************************************************
Date         : 2020-02-04
Author       : Xiaojun Zou
Harman International Industries, Incorporated

Description  : there are following 2 updates: 1. fixing the crash issue after the MTP device disconnected; 2. add building version log;
ChangeList   : 5501268
*******************************************************************************

diff --git a/src/simple-mtpfs-main.cpp b/src/simple-mtpfs-main.cpp
index fba38da..a2be1e8 100755
--- a/src/simple-mtpfs-main.cpp
+++ b/src/simple-mtpfs-main.cpp
@@ -26,6 +26,8 @@ int main(int argc, char **argv)
 
     SMTPFileSystem *filesystem = SMTPFileSystem::instance();
 
+    //  log the build time
+    std::cout << "the building version of simple-mtpfs: " << getVersion() << std::endl;
     if (!filesystem->parseOptions(argc, argv)) {
         std::cout << "Wrong usage! See `" << smtpfs_basename(argv[0])
             << " -h' for details\n";
diff --git a/src/simple-mtpfs-mtp-device.cpp b/src/simple-mtpfs-mtp-device.cpp
index 87b9e69..6a66a97 100755
--- a/src/simple-mtpfs-mtp-device.cpp
+++ b/src/simple-mtpfs-mtp-device.cpp
@@ -232,8 +232,9 @@ bool MTPDevice::connect_priv(int dev_no, const std::string &dev_file)
     if(m_rbuf)
     {
         RingBuffer_Free(m_rbuf);
+		m_rbuf = NULL;
     }
-    else
+    //	else
     {
         m_rbuf = RingBuffer_Malloc(RINGBUF_SIZE);
     }
@@ -286,7 +287,11 @@ void MTPDevice::disconnect()
 	criticalEnter1();
     LIBMTP_Release_Device(m_device);
 	criticalLeave1();
-	RingBuffer_Free(m_rbuf);
+	if (m_rbuf)
+	{
+		RingBuffer_Free(m_rbuf);
+		m_rbuf = NULL;
+	}
     m_device = nullptr;
     g_device = m_device;
     mUUID = "";
@@ -328,7 +333,7 @@ bool MTPDevice::enumStorages()
 
 const TypeDir *MTPDevice::dirFetchContentFromCache(std::string path)
 {
-    //LogInfo("path:%s", path.c_str());
+    //LogInfo("MTPDevice::dirFetchContentFromCache: path:%s", path.c_str());
 
     if (false == m_root_dir.isFetched()) {
         //int count = 0;
diff --git a/src/simple-mtpfs-util.cpp b/src/simple-mtpfs-util.cpp
index e2ffcbc..cc56969 100755
--- a/src/simple-mtpfs-util.cpp
+++ b/src/simple-mtpfs-util.cpp
@@ -49,6 +49,24 @@ bool StreamHelper::s_enabled = false;
 int StreamHelper::s_stdout = -1;
 int StreamHelper::s_stderr = -1;
 
+std::string getVersion()
+{
+   static char ver[8];
+   struct tm t;
+   const std::string buildTime = __DATE__ " " __TIME__;
+   std::stringstream ss;
+
+   memset(ver, 0, sizeof(ver));
+   memset(&t, 0, sizeof(t));
+
+   strptime(buildTime.c_str(), "%b %d %Y %T", &t);
+   strftime(ver, sizeof(ver), "%g%V%u", &t);
+
+   ss << "[" << ver << "]" << "[" << buildTime << "]";
+
+   return ss.str();
+}
+
 void StreamHelper::on()
 {
     if (Logger::isEnableLibmtpLog())
diff --git a/src/simple-mtpfs-util.h b/src/simple-mtpfs-util.h
index 351c3ad..c08b724 100755
--- a/src/simple-mtpfs-util.h
+++ b/src/simple-mtpfs-util.h
@@ -36,6 +36,8 @@
 // typedef unsigned int uint32_t;
 // typedef unsigned long long uint64_t;
 
+std::string getVersion();
+
 class StreamHelper
 {
 public:
